// scrapfly-sdk/script/main.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.log = exports.ScreenshotResult = exports.ScrapeResult = exports.ExtractionResult = exports.errors = exports.ExtractionConfig = exports.ScreenshotConfig = exports.ScreenshotOptions = exports.ScreenshotFormat = exports.ScreenshotFlags = exports.ScrapeConfig = exports.ScrapeFormatOptions = exports.ScrapeFormat = exports.ScrapflyClient = void 0;
var client_js_1 = require("./client.js");
Object.defineProperty(exports, "ScrapflyClient", { enumerable: true, get: function () { return client_js_1.ScrapflyClient; } });
var scrapeconfig_js_1 = require("./scrapeconfig.js");
Object.defineProperty(exports, "ScrapeFormat", { enumerable: true, get: function () { return scrapeconfig_js_1.Format; } });
Object.defineProperty(exports, "ScrapeFormatOptions", { enumerable: true, get: function () { return scrapeconfig_js_1.FormatOption; } });
Object.defineProperty(exports, "ScrapeConfig", { enumerable: true, get: function () { return scrapeconfig_js_1.ScrapeConfig; } });
Object.defineProperty(exports, "ScreenshotFlags", { enumerable: true, get: function () { return scrapeconfig_js_1.ScreenshotFlags; } });
var screenshotconfig_js_1 = require("./screenshotconfig.js");
Object.defineProperty(exports, "ScreenshotFormat", { enumerable: true, get: function () { return screenshotconfig_js_1.Format; } });
Object.defineProperty(exports, "ScreenshotOptions", { enumerable: true, get: function () { return screenshotconfig_js_1.Options; } });
Object.defineProperty(exports, "ScreenshotConfig", { enumerable: true, get: function () { return screenshotconfig_js_1.ScreenshotConfig; } });
var extractionconfig_js_1 = require("./extractionconfig.js");
Object.defineProperty(exports, "ExtractionConfig", { enumerable: true, get: function () { return extractionconfig_js_1.ExtractionConfig; } });
exports.errors = __importStar(require("./errors.js"));
var result_js_1 = require("./result.js");
Object.defineProperty(exports, "ExtractionResult", { enumerable: true, get: function () { return result_js_1.ExtractionResult; } });
Object.defineProperty(exports, "ScrapeResult", { enumerable: true, get: function () { return result_js_1.ScrapeResult; } });
Object.defineProperty(exports, "ScreenshotResult", { enumerable: true, get: function () { return result_js_1.ScreenshotResult; } });
var logger_js_1 = require("./logger.js");
Object.defineProperty(exports, "log", { enumerable: true, get: function () { return logger_js_1.log; } });
